pkg/ui/src/util/loader.tsx

```diff
@@ -0,0 +1,85 @@
+import React from "react";
+
+import spinner from "assets/spinner.gif";
+
+interface LoaderProps<T> {
+  load: () => Promise<T>;
+  children: (data: T, handleRefresh: () => void) => JSX.Element;
+}
+
+interface LoaderState<T> {
+  state: string;
+  data?: T;
+  error?: any;
+}
+
+/**
+ * Loader handles data loading. Give it a data loading function and a render
+ * function (`load` and `children`, respectively), and it will:
+ * - render a spinner while data is loading.
+ * - render an error when the API request returns one.
+ * - render the data using the provided function when the API call returns.
+ *
+ * The render function is also passsed a `handleRefresh` function, which can
+ * be used as an onClick handler to make a refresh button.
+ */
+class Loader<T> extends React.Component<LoaderProps<T>, LoaderState<T>> {
```

Comment:
note to reviewers (@couchand, @mrtracy): added this as an experimental alternative to our usual Redux/`CachedDataReducer` song and dance. I think it worked out pretty well, but I'm happy to take this out and use `CachedDataReducer` in this PR for the sake of consistency, until we decide on a more comprehensive data loading refactor (i.e. @mrtracy's sagas+HOC thing)